Ananyamukh6/total variation

ananyamukh6 opened the pull request:
Implements total variation

Also adds an example using the added total variation function to denoise an image

ducha-aiki commented on examples/total_variation_denoising.py:
```diff
@@ -0,0 +1,73 @@
+import kornia
+import cv2
+import numpy as np
+from torch.autograd import Variable
```
Seems not used here

ananyamukh6 replied:
Removed extra import

ducha-aiki commented on examples/total_variation_denoising.py:
```diff
@@ -0,0 +1,73 @@
+
+# convert to torch tensor
+noisy_image: torch.tensor = kornia.image_to_tensor(img).squeeze()  # CxHxW
+noisy_image
```
remove

ananyamukh6 replied:
Removed

ducha-aiki commented on examples/total_variation_denoising.py:
```diff
@@ -0,0 +1,73 @@
+    optimizer.zero_grad()
+    loss = tv_denoiser()
+    if i % 25 == 0:
+        print("Loss in iteration ", i, " of ", num_iters, ": ", loss.detach().numpy())
```
loss.detach().numpy() -> loss.item() 

Nicer output, no need of array

ananyamukh6 replied:
Thanks for the suggestion. Used `item`

ducha-aiki commented on kornia/losses/total_variation.py:
```diff
@@ -0,0 +1,44 @@
+        pixel_dif2 = img[:, :, :, 1:] - img[:, :, :, :-1]
+        reduce_axes = (1, 2, 3)
+    else:
+        raise ValueError("Expected input tensor to be of rank 3 or 4, but got " + str(len(img_shape)))
```
I believe, it is not "rank", but ndim. 

"The total number of contravariant and covariant indices of a tensor. The rank R of a tensor is independent of the number of dimensions N of the underlying space."

http://mathworld.wolfram.com/TensorRank.html

edgarriba replied:
This is not needed using ellipsis. We can handle (*,H,W)
```python
pixel_dif1 = img[..., 1:, :] - img[..., :-1, :]
pixel_dif2 = img[..., :, 1:] - img[..., :, :-1]
```

edgarriba replied:
Besides I believe it can be done with a conv2d using a filter of [1, -1] including padding so that we can recover the full resolution image

edgarriba replied:
Whatever is faster is fine to me

ananyamukh6 replied:
@edgarriba 
I did think of using convolution. But the example for denoising I have ran in ~9 secs with conv (while with the current diff implementation it ran in ~2 secs). (ran on CPU, not GPU). The time diff was quite surprising to me!

here is the code for total variation implemented in terms of convolution. Let me know if you see any scope of optimizing the conv based implementation and I can try it:
```
class TotalVariation(nn.Module):
    def __init__(self) -> None:
        super(TotalVariation, self).__init__()
        self.conv_3d_1 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=(1,2), bias=False)
        self.conv_3d_1.weight = torch.nn.Parameter(data=torch.tensor([[[[1.0,-1.0]]]]), requires_grad=False)
        self.conv_3d_2 = torch.nn.Conv2d(in_channels=1, out_channels=1, kernel_size=(2,1), bias=False)
        self.conv_3d_2.weight = torch.nn.Parameter(data=torch.tensor([[[[ 1.0],[-1.0]]]]), requires_grad=False)

    def forward(  # type: ignore
            self, img) -> torch.Tensor:
        #return total_variation(img)
        if not torch.is_tensor(img):
            raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
        img_shape = img.shape
        if not (len(img_shape) == 3 or len(img_shape) == 4):
            raise ValueError("Expected input tensor to be of ndim 3 or 4, but got " + str(len(img_shape)))
        if len(img_shape) == 3:
            # unsqueeze introduces a batch dimension
            img = img.unsqueeze(0)
        num_channels = img_shape[1]
        channel_separated = torch.chunk(img, chunks=num_channels, dim=1)
        # channel_separated is C tensors of size N x 1 x H x W

        to_be_added = []
        for channel in channel_separated:
            to_be_added.append(self.conv_3d_1(channel).abs().sum(dim=(1, 2, 3)))
            to_be_added.append(self.conv_3d_2(channel).abs().sum(dim=(1, 2, 3)))
        total : torch.Tensor = torch.stack(to_be_added).sum(dim=0)
        if len(img_shape) == 3:
            # get rid of the batch dimension
            total = torch.squeeze(total)
        return total


def total_variation(img: torch.Tensor) -> torch.Tensor:
    return TotalVariation()(img)
```

edgarriba replied:
filter2D already implements that. Give it a shot and check the performance
https://kornia.readthedocs.io/en/latest/filters.html#kornia.filters.filter2D

ananyamukh6 replied:
Tried `kornia.filter.filter2D`, but it is still slow. I think that its just conv in this case is slower than subtraction. I suggest we stick to the current difference based implementation

Note I tried on CPU
```
class TotalVariation(nn.Module):
    def __init__(self) -> None:
        super(TotalVariation, self).__init__()
        self.weight1 = torch.nn.Parameter(data=torch.tensor([[[1.0,-1.0]]]), requires_grad=False)
        self.weight2 = torch.nn.Parameter(data=torch.tensor([[[ 1.0],[-1.0]]]), requires_grad=False)

    def forward(  # type: ignore
            self, img) -> torch.Tensor:
        if not torch.is_tensor(img):
            raise TypeError(f"Input type is not a torch.Tensor. Got {type(img)}")
        img_shape = img.shape
        if not (len(img_shape) == 3 or len(img_shape) == 4):
            raise ValueError("Expected input tensor to be of ndim 3 or 4, but got " + str(len(img_shape)))
        if len(img_shape) == 3:
            # unsqueeze introduces a batch dimension
            img = img.unsqueeze(0)
        num_channels = img_shape[1]
        channel_separated = torch.chunk(img, chunks=num_channels, dim=1)
        # channel_separated is C tensors of size N x 1 x H x W

        to_be_added = []
        for channel in channel_separated:
            to_be_added.append(kornia.filters.filter2D(channel, self.weight1).abs().sum(dim=(1, 2, 3)))

            to_be_added.append(kornia.filters.filter2D(channel, self.weight2).abs().sum(dim=(1, 2, 3)))
        total : torch.Tensor = torch.stack(to_be_added).sum(dim=0)
        if len(img_shape) == 3:
            # get rid of the batch dimension
            total = torch.squeeze(total)
        return total
```

edgarriba replied:
Great

ducha-aiki commented on test/test_losses.py:
```diff
@@ -265,3 +267,58 @@ def test_kl_div_loss_2d(self, input, target, expected):
+        expected = torch.tensor(np.sum(np.abs(pixel_dif1)) + np.sum(np.abs(pixel_dif2)))
+        assert_allclose(actual, expected)
+
+    # Total variation for 3D tensors
```
4D

ananyamukh6 replied:
Fixed comment

dvd42 commented on examples/total_variation_denoising.py:
```diff
@@ -0,0 +1,71 @@
+    optimizer.zero_grad()
+    loss = tv_denoiser()
+    if i%25 == 0:
+        print("Loss in iteration ", i, " of ", num_iters, ": ", loss.item())
```
please use the `'{}'.format()` format for printing. Something like:
`print("Loss in iteration {} of {}: {:.3f}".format(i, num_iters, loss.item()))`
It's much cleaner

ananyamukh6 replied:
Good idea! Done.

edgarriba replied:
not for this PR, but I recently started using f-string format, which is really cool !
https://realpython.com/python-f-strings/

ananyamukh6 commented on test/test_losses.py:
```diff
@@ -265,3 +267,58 @@ def test_kl_div_loss_2d(self, input, target, expected):
+        input_np = input.numpy()
+        pixel_dif1 = input_np[:, :, 1:, :] - input_np[:, :, :-1, :]
+        pixel_dif2 = input_np[:, :, :, 1:] - input_np[:, :, :, :-1]
+        expected = torch.tensor(np.sum(np.abs(pixel_dif1), axis=(1, 2, 3)) + np.sum(np.abs(pixel_dif2), axis=(1, 2, 3)))
```
The TotalVariation code does these ops in torch terms. This is a test, hence I am using numpy to compute the expected values. Using torch to compute the values expected of the torch function is slightly circular.

dvd42 replied:
I see your point, however it would be nice if the tests would have the least amount of dependencies as possible. Image numpy changes how it handles axis in the future, then the tests will fail when the code is correct. Of course, it is extremely unlikely that numpy changes how it handles axis in the future but you see my point. It is nice that the code  uses only torch, for consistency sake.

One way you could do these tests is by providing an input for which you know the expected output see [this example](https://github.com/arraiyopensource/kornia/blob/d97f76fe77ab19744a7def88e5dcefc71ddbae0e/test/color/test_hsv.py#L14). 

P.S: Thank you for your pacience, I am aware I am being a bit picky with these changes.

bhack replied:
The TF version use a double approach for the tests:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_test.py#L4190

dvd42 replied:
A double approach would also be valid althought I dont know if we necessarily need/want it

bhack replied:
See also this comment https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/image_ops_test.py#L4284 and the follow-up handmade test.

edgarriba replied:
vote for avoid using numpy

ananyamukh6 replied:
Removed numpy dependency

ananyamukh6 commented:
Addressed comments by reviewers. Please re-review

Also I see codecov fail (red crosses on them). Can someone familiar with the repo please tell me if that is expected, and if not what I can do to correct it.

edgarriba commented on kornia/losses/total_variation.py:
```diff
@@ -0,0 +1,40 @@
+    if len(img_shape) == 3 or len(img_shape) == 4:
+        pixel_dif1 = img[..., 1:, :] - img[..., :-1, :]
+        pixel_dif2 = img[..., :, 1:] - img[..., :, :-1]
+        reduce_axes = (0, 1, 2) if len(img_shape) == 3 else (1, 2, 3)
```
yes, plus ellipsis

ananyamukh6 commented:
Addressed all comments. Please re-review.

I notice that CI is failing, but its failing at the "install dependencies" stage. Is CI down?

ananyamukh6 commented:
> LGTM. Awesome contribution. Will merge once we fix the CI build issue

Thanks!
CI is passing now

edgarriba commented:
@ananyamukh6 you made it ! thanks for the contribution. Anything else in your mind to be implemented ?
